File: leader-zmq/gravity.py
import time
import threading
import numpy as np
import csv
from datetime import datetime
from dynamixel_sdk import (
    PortHandler,
    PacketHandler,
    GroupSyncRead,
    GroupSyncWrite,
)
import pinocchio as pin
from pinocchio.robot_wrapper import RobotWrapper

# === Paths and Model ===
URDF_PATH = "robot.urdf"
MESH_DIR = ["meshes"]

# Load robot model (Pinocchio)
robot = RobotWrapper.BuildFromURDF(URDF_PATH, MESH_DIR)
model = robot.model
data = robot.data

# === Dynamixel Constants ===
DEVICENAME = "/dev/ttyUSB0"
BAUDRATE = 4000000
PROTOCOL_VERSION = 2.0

ADDR_PRESENT_POSITION = 132
LEN_PRESENT_POSITION = 4
ADDR_GOAL_CURRENT = 102
LEN_GOAL_CURRENT = 2
ADDR_TORQUE_ENABLE = 64
ADDR_OPERATING_MODE = 11

MOTOR_IDS = list(range(7))
DXL_RESOLUTION = 4096

# Torque constants and current LSB per joint (adjust for your model)
TORQUE_CONSTANTS = np.array([1.1] * 7)  # [Nm/A] for each joint
DXL_CURRENT_UNIT = 2.69e-3  # [A/LSB] for XL430, XM430. Change if needed!

# === Control Parameters ===
STIFFNESS = np.diag([6, 40, 30, 9.0, 0.8, 0.8, 0.0])
DAMPING = np.diag([2.0, 2.0, 1.2, 1.0, 0.8, 0.8, 0.1])
FEEDFORWARD_GAIN = np.diag([6, -10, 10, 4, 2, 2, 0.1])
VEL_THRESHOLD = 0.01
STILL_CYCLES = 30
CONTROL_FREQ = 50
DT = 1.0 / CONTROL_FREQ

# === Initialize SDK ===
portHandler = PortHandler(DEVICENAME)
packetHandler = PacketHandler(PROTOCOL_VERSION)
assert portHandler.openPort()
portHandler.setBaudRate(BAUDRATE)

groupRead = GroupSyncRead(portHandler, packetHandler, ADDR_PRESENT_POSITION, LEN_PRESENT_POSITION)
groupWrite = GroupSyncWrite(portHandler, packetHandler, ADDR_GOAL_CURRENT, LEN_GOAL_CURRENT)

for dxl_id in MOTOR_IDS:
    packetHandler.write1ByteTxRx(portHandler, dxl_id, ADDR_TORQUE_ENABLE, 0)
    packetHandler.write1ByteTxRx(portHandler, dxl_id, ADDR_OPERATING_MODE, 0)  # current mode
    packetHandler.write1ByteTxRx(portHandler, dxl_id, ADDR_TORQUE_ENABLE, 1)
    groupRead.addParam(dxl_id)

def raw_to_rad(raw, center):
    # Convert Dynamixel raw position to radians (accounting for captured zero offset)
    return (raw - center) * (2 * np.pi / DXL_RESOLUTION)

def compute_gravity_torque(q):
    # Pinocchio expects shape (7,), radians
    pin.computeGeneralizedGravity(model, data, q)
    return data.g.copy()

def send_currents(currents):
    groupWrite.clearParam()
    for i, dxl_id in enumerate(MOTOR_IDS):
        # Convert Nm to Amps using torque constant, then Amps to LSB using DXL_CURRENT_UNIT
        amps = currents[i] / TORQUE_CONSTANTS[i]
        current_lsb = int(np.clip(amps / DXL_CURRENT_UNIT, -2047, 2047))
        param = [current_lsb & 0xFF, (current_lsb >> 8) & 0xFF]
        groupWrite.addParam(dxl_id, param)
    groupWrite.txPacket()

raw_zero = {}

def zero_pose_capture():
    global raw_zero
    print("Waiting for joint data to capture zero pose...")
    while True:
        groupRead.txRxPacket()
        if all(groupRead.isAvailable(dxl_id, ADDR_PRESENT_POSITION, LEN_PRESENT_POSITION) for dxl_id in MOTOR_IDS):
            raw_zero = {
                dxl_id: groupRead.getData(dxl_id, ADDR_PRESENT_POSITION, LEN_PRESENT_POSITION)
                for dxl_id in MOTOR_IDS
            }
            print("Zero pose captured.")
            break
        time.sleep(0.1)

class ControlLoop:
    def __init__(self):
        self.q_target = np.zeros(7)
        self.q_prev = np.zeros(7)
        self.hold_enabled = False

    def hold_position(self, q):
        self.q_target = q.copy()
        print("🔵 HOLD: Holding compliant pose at current position.")

    def run(self):
        still_counter = 0
        debug_counter = 0
        log_file = open("impedance_log.csv", "w", newline="")
        csv_writer = csv.writer(log_file)
        csv_writer.writerow(
            ["time"] +
            [f"q{i}" for i in range(7)] +
            [f"q_err{i}" for i in range(7)] +
            [f"dq{i}" for i in range(7)] +
            [f"tau{i}" for i in range(7)]
        )

        while True:
            start_time = time.time()
            groupRead.txRxPacket()
            q_raw = [groupRead.getData(dxl_id, ADDR_PRESENT_POSITION, LEN_PRESENT_POSITION) for dxl_id in MOTOR_IDS]
            q = np.array([raw_to_rad(q_raw[i], raw_zero[MOTOR_IDS[i]]) for i in range(7)])
            dq = (q - self.q_prev) / DT
            self.q_prev = q.copy()

            if not self.hold_enabled:
                if np.linalg.norm(dq) < VEL_THRESHOLD:
                    still_counter += 1
                else:
                    still_counter = 0
                if still_counter >= STILL_CYCLES:
                    self.q_target = q.copy()
                    still_counter = 0
                    print("🟢 Updated floating hold pose (auto)")

            if self.hold_enabled:
                self.hold_position(q)
                self.hold_enabled = False

            q_error = np.clip(q - self.q_target, -0.5, 0.5)
            tau_g = compute_gravity_torque(q)
            tau_imp = -STIFFNESS @ q_error
            tau_damp = -DAMPING @ dq
            tau_ff = FEEDFORWARD_GAIN @ dq
            tau = tau_g + tau_imp + tau_damp + tau_ff

            send_currents(tau)

            now = time.time()
            csv_writer.writerow([now] + list(q) + list(q_error) + list(dq) + list(tau))

            debug_counter += 1
            if debug_counter % 25 == 0:
                print(f"DEBUG q: {q.round(3)} | q_err: {q_error.round(3)} | dq: {dq.round(3)} | tau_g: {tau_g.round(3)}")

            time.sleep(max(0.0, DT - (time.time() - start_time)))

def input_thread(loop_obj):
    while True:
        cmd = input("Type 'h' + Enter to HOLD, 'f' + Enter to FLOAT: ").strip().lower()
        if cmd == "h":
            loop_obj.hold_enabled = True
        elif cmd == "f":
            print("🟠 FLOAT: Floating mode resumed.")
        else:
            print("Unknown command.")

if __name__ == "__main__":
    zero_pose_capture()
    print("Starting compliant impedance control with gravity compensation and logging...")
    loop = ControlLoop()
    control_thread = threading.Thread(target=loop.run)
    control_thread.daemon = True
    control_thread.start()
    input_thread(loop)

File: leader-zmq/gravity.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <chrono>
#include <cmath>
#include <Eigen/Dense>
#include "dynamixel_sdk/dynamixel_sdk.h"

using namespace dynamixel;
using namespace Eigen;

constexpr char DEVICENAME[] = "/dev/ttyUSB0";
constexpr int BAUDRATE = 4000000;
constexpr float TORQUE_CONSTANT = 1.1f;
constexpr float DXL_RESOLUTION = 4096.0f;
constexpr float PI = 3.14159265358979323846f;
constexpr int NUM_JOINTS = 7;
constexpr float DT = 1.0f / 50.0f;

const uint16_t ADDR_PRESENT_POSITION = 132;
const uint16_t ADDR_GOAL_CURRENT = 102;
const uint8_t ADDR_OPERATING_MODE = 11;
const uint8_t ADDR_TORQUE_ENABLE = 64;

PortHandler *portHandler;
PacketHandler *packetHandler;
GroupSyncRead *groupRead;
GroupSyncWrite *groupWrite;
std::vector<uint8_t> MOTOR_IDS = {0, 1, 2, 3, 4, 5, 6};
std::map<uint8_t, int32_t> raw_zero;

Matrix<float, NUM_JOINTS, 1> rawToRad(const std::vector<int32_t>& raw) {
    Matrix<float, NUM_JOINTS, 1> q;
    for (int i = 0; i < NUM_JOINTS; ++i)
        q(i) = (raw[i] - raw_zero[MOTOR_IDS[i]]) * (2 * PI / DXL_RESOLUTION);
    return q;
}

void sendCurrents(const VectorXf& tau) {
    groupWrite->clearParam();
    for (int i = 0; i < NUM_JOINTS; ++i) {
        float amps = tau(i) / TORQUE_CONSTANT;
        int current = static_cast<int>(std::round(amps));
        current = std::min(std::max(current, -2047), 2047);
        uint8_t param[2] = { static_cast<uint8_t>(current & 0xFF), static_cast<uint8_t>((current >> 8) & 0xFF) };
        groupWrite->addParam(MOTOR_IDS[i], param);
    }
    groupWrite->txPacket();
}

void configureMotors() {
    for (auto id : MOTOR_IDS) {
        int dxl_comm_result;
        dxl_comm_result = packetHandler->write1ByteTxRx(portHandler, id, ADDR_TORQUE_ENABLE, 0);
        dxl_comm_result |= packetHandler->write1ByteTxRx(portHandler, id, ADDR_OPERATING_MODE, 0);
        dxl_comm_result |= packetHandler->write1ByteTxRx(portHandler, id, ADDR_TORQUE_ENABLE, 1);
        if (dxl_comm_result != COMM_SUCCESS) {
            std::cerr << "Failed to configure motor ID " << static_cast<int>(id) << std::endl;
        }
        groupRead->addParam(id);
    }
}

int main() {
    portHandler = PortHandler::getPortHandler(DEVICENAME);
    packetHandler = PacketHandler::getPacketHandler(2.0);
    groupRead = new GroupSyncRead(portHandler, packetHandler, ADDR_PRESENT_POSITION, 4);
    groupWrite = new GroupSyncWrite(portHandler, packetHandler, ADDR_GOAL_CURRENT, 2);

    portHandler->openPort();
    portHandler->setBaudRate(BAUDRATE);

    configureMotors();

    std::cout << "Capturing zero pose...\n";
    while (true) {
        groupRead->txRxPacket();
        bool all_available = true;
        for (auto id : MOTOR_IDS) {
            if (!groupRead->isAvailable(id, ADDR_PRESENT_POSITION, 4)) {
                all_available = false;
                break;
            }
        }
        if (all_available) {
            for (auto id : MOTOR_IDS) {
                raw_zero[id] = groupRead->getData(id, ADDR_PRESENT_POSITION, 4);
            }
            break;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    std::cout << "Zero pose captured.\n";

    VectorXf q_prev = VectorXf::Zero(NUM_JOINTS);
    VectorXf q_target = VectorXf::Zero(NUM_JOINTS);
    const Matrix<float, NUM_JOINTS, NUM_JOINTS> STIFFNESS = Vector<float, NUM_JOINTS>({10, 30, 20, 4, 2, 0.5f, 0.0f}).asDiagonal();
    const Matrix<float, NUM_JOINTS, NUM_JOINTS> DAMPING = Vector<float, NUM_JOINTS>({2, 5, 2.2f, 2, 0.8f, 0.8f, 0.1f}).asDiagonal();
    const Matrix<float, NUM_JOINTS, NUM_JOINTS> FEEDFORWARD = Vector<float, NUM_JOINTS>({8, 10, 10, 4, 1, 1, 0.1f}).asDiagonal();

    int still_counter = 0;
    const float vel_thresh = 0.01f;

    while (true) {
        groupRead->txRxPacket();
        std::vector<int32_t> raw;
        for (auto id : MOTOR_IDS)
            raw.push_back(groupRead->getData(id, ADDR_PRESENT_POSITION, 4));

        VectorXf q = rawToRad(raw);
        VectorXf dq = (q - q_prev) / DT;
        q_prev = q;

        if (dq.norm() < vel_thresh)
            still_counter++;
        else
            still_counter = 0;

        if (still_counter > 30) {
            q_target = q;
            still_counter = 0;
            std::cout << "Updated floating hold pose.\n";
        }

        VectorXf q_error = (q - q_target).cwiseMax(-0.5f).cwiseMin(0.5f);
        VectorXf tau = -STIFFNESS * q_error - DAMPING * dq + FEEDFORWARD * dq;
        sendCurrents(tau);

        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(DT * 1000)));
    }

    return 0;
}